### DISCUSSION TITLE
Foreign Key as Primary Key

### DISCUSSION THREAD

**Discussion by NDatis** [api-platform/api-platform#1782] [Q&A] [answered]
Hi !

When using a [foreign key as primary key](https://www.doctrine-project.org/projects/doctrine-orm/en/latest/tutorials/composite-primary-keys.html#identity-through-foreign-entities)
Is there a way to specify API Platform to use the ID value as the route parameter ? Because the route actually expects a campaign instance as parameter

Any help would be very appreciated !

**Comment by soyuka** [comment, marked as answer]
try https://github.com/api-platform/core/blob/main/docs/adr/0001-resource-identifiers.md (look for `/users/{keya}/{keyb}`) on 2.6

**Reply by NDatis** [reply]
Hi, first thanks a lot for your work ! 

Is the `/companies/{companyId}/users/{id}` supposed to be working on 2.6 ? 

Because I get an "Unknown property "path" on annotation" error, maybe the question is stupid, but I just can't figure out how to make a "subresource" POST route manually

Would be really grateful for your help !

**Reply by soyuka** [reply]
try adding a [custom item operation](https://api-platform.com/docs/core/operations/#enabling-and-disabling-operations) on User, the path `/companies/{companyId}/users/{id}`, method `POST` and `identifiers` to `['companyId' => [Company::class, 'id'], 'id' => [User::class, 'id']]` .

**Reply by NDatis** [reply]
Thanks again for your help, but when I define other key than simple auto-incrementing id, the statement still selects by 'id' and not the key I defined

Is there a solution for that ?

**Reply by NDatis** [reply]
For example : 

```
/**
 * @ApiResource(
 *   collectionOperations={},
 *   itemOperations={
 *         "post_respondent"={
 *             "method" = "POST",
 *             "path"="/campaigns/surveys/{survey}/respondents",
 *             "identifiers"={"survey": {Survey::class, "campaign"}},
 *             "controller"=SaveRespondent::class
 *         }
 *     }
 * )
 */
class Respondent
{
    /**
     * @ORM\Id
     * @ORM\GeneratedValue
     * @ORM\Column(type="integer")
     */
    private $id;

    /**
     * @ORM\ManyToOne(targetEntity=Survey::class, inversedBy="respondents")
     */
    private $survey;
}
```

```
/**
 * @ApiResource()
 */
class Survey
{
    /**
     * @ORM\Id
     * @ORM\OneToOne(targetEntity=Campaign::class, inversedBy="survey")
     * @ORM\JoinColumn(name="campaign_uuid", referencedColumnName="uuid", nullable=false)
     */
    private $campaign;
}
```

```
/**
 * @ApiResource()
 */
class Campaign
{
    /**
     * @ORM\Id
     * @ORM\Column(type="uuid", unique=true)
     */
    private $uuid;

    /**
     * @ORM\OneToOne(targetEntity=Survey::class, mappedBy="campaign")
     */
    private $survey;
}

```
The following entities configuration gives this error :

> An exception occurred while executing 'SELECT ... WHERE _.survey_id = ?' with params [\"206b00da-4218-4b73-8c2e-8ff138aada92\"]:\n\nSQLSTATE[22P02]: Invalid text representation: 7 ERROR:  invalid input syntax for type integer: \"206b00da-4218-4b73-8c2e-8ff138aada92\""
